### PR TITLE
[DoRA] add dora

### PR DESCRIPTION
**Pull request by JunnYu**
### PR types

New features

### PR changes

APIs

### Description

image
- Paper: https://arxiv.org/abs/2402.09353
- DoRA decomposes the pre-trained weight into two components, magnitude and direction, for fine-tuning, specifically employing LoRA for directional updates to efficiently minimize the number of trainable parameters. By employing DoRA, we enhance both the learning capacity and training stability of LoRA while avoiding any additional inference overhead. DoRA consistently outperforms LoRA on fine-tuning LLaMA, LLaVA, and VL-BART on various downstream tasks, such as commonsense reasoning, visual instruction tuning, and image/video-text understanding
- DoRA hyperparameters settings
[NOTE that] 💡 While fine-tuning with DoRA by utilizing the configuration of LoRA can already achieve better results most of the time, achieving optimal performance compared to LoRA still requires adjustments to the hyperparameters.
We suggest starting with a slightly lower learning rate than that of LoRA, and users may also experiment with varying lora dropout ratios.
User may also start with half of the rank of the LoRA configuration which oftentime can already results in comparable or even superior accuracy compared to that of LoRA.
- Usage:
`python finetune_generation.py ./llama/lora_argument.json --use_dora True`

Reference:
- https://github.com/huggingface/peft/tree/main/src/peft/tuners/lora
- https://github.com/nbasyl/DoRA
- https://arxiv.org/abs/2402.09353

### CODE REVIEW

File: paddlenlp/peft/lora/lora_layers.py [hunk 1]
```diff
@@ -69,26 +70,63 @@ def __init__(
...
     def forward(self, input: paddle.Tensor, *args, **kwargs):
         result = F.linear(x=input, weight=self.weight, bias=self.bias, name=self.name)
         if not self.merged:
-            result += (self.lora_dropout(input) @ self.lora_A @ self.lora_B) * self.scaling
+            if self.use_dora:
+                input = self.lora_dropout(input)
+                dora_weight = self.weight + self.lora_A @ self.lora_B * self.scaling
+                weight_norm = dora_weight.norm(p=2, axis=0, keepdim=True).detach()
+                mag_norm_scale = (self.lora_magnitude / weight_norm).unsqueeze(0)
+                result_dora = (mag_norm_scale - 1) * (input @ self.weight) + mag_norm_scale * (
```
Comment (gongel) [inline paddlenlp/peft/lora/lora_layers.py]: 这里能复用吗？mag_norm_scale * input @ dora_weight - input @ self.weight(这里从result里面取)

Comment (JunnYu) [reply paddlenlp/peft/lora/lora_layers.py]: 有dropout，所以结果不一样，只有当dropout=0的时候可以复用
